### ISSUE TITLE
Add a better hash function.

### ISSUE THREAD

**Issue by orlp** [isaacg1/pyth#65]
The built-in `hash` kind of sucks, we should probably build a better hash function for `.h`.

**Comment by isaacg1** [comment]
Any particular complaints?

**Comment by orlp** [comment]
@isaacg1 The hash function is very weak (has many collisions) and is not constant between executions. For example this program has a different output if run multiple times:

    .h\0

**Comment by Maltysen** [comment]
SHA is a good idea, but the simple return type of an int in the default hash was good, so we should implement that. Also, to get around all the `not hashable` and make it general we can apply sha to the `.__repr__` of the object.

**Comment by orlp** [comment]
@Maltysen  I disagree with the simple return on int. You're only hashing ints in Pyth if you want to uniformly distribute them.

**Comment by isaacg1** [comment]
I'm not sure if this is what Maltysen meant, but I like returning ints from `.h`, as opposed to a string.

**Comment by orlp** [comment]
@isaacg1 I definitely think `.h` should return int, I thought @Maltysen was talking about the fact that for integers the default Python hash is the identity function.

**Comment by Maltysen** [comment]
I did mean what @isaacg1 was saying and agree that it being an identity for ints is useless.

**Comment by willsherwood** [comment]
How is the identity function for hashes on ints useless for int hashing?

-------- Original message --------

**Comment by orlp** [comment]
@willsherwood  I expect the output of a hash to be uniform over its output range regardless of the distribution of input. Identity on integers does not preserve this expectation.

**Comment by orlp** [comment]
I've implemented a better hash function. Since hash functions in Pyth won't be used to compare elements for equality or build data structures I've made the decision to group different types:

- `bool` and `int`
- `str`
- `tuple`, `list`, `set`, `dict`

`bool` gets transformed to `0`/`1`. `set` is hashed as if it's an ordered sequence, `dict` is hashed as if it's an ordered sequence of `(key, value)` tuples.
